[4.3] Fix submit button focus on frontend MFA captive login page when using Webauthn

richard67 opened the pull request:
Pull Request for https://github.com/joomla/joomla-cms/pull/40330#issuecomment-1530282790 .

### Summary of Changes

When working on my pull request (PR) #40330 I've noticed that when using MFA at the site (frontend) login and the currently active MFA method is Webauthn, the javascript "com_users/js/two-factor-focus" is not loaded, and so the submit button is not activated, but when doing the same at the admin (backend) login, the script is loaded so the button is activated.

The reason for this is the `if` condition for loading the javascript in the site captive login here: https://github.com/richard67/joomla-cms/blob/4.3-dev/components/com_users/tmpl/captive/default.php#L25-L28

We do not have this condition for the admin captive login page: https://github.com/richard67/joomla-cms/blob/4.3-dev/administrator/components/com_users/tmpl/captive/default.php#L25-L26

This PR here removes this condition in file `components/com_users/tmpl/captive/default.php` and does a small code style fix in addition in that file.

In addition it fixes the `two-factor-focus.es6.js` so that it works with the above change. This basically means a revert of my PR #40330 , which has fixed the issue handled by that PR but broke the thing handled with this PR here, and replaces that wrong fix by the right one, using optional chaining as recommended by @nikosdion during review of this PR here, thanks for that.

### Testing Instructions

First test without this PR applied to reproduce the issue.

After applying the changes from this PR, run `npm ci` or `npm run build:js` when on a development environment.

When on a regular installation where you have no npm, use the update package or the custom update URL created by drone for this PR.

After having applied the changes, clear your browser cache to make sure you are not testing the old, unmodified javascript.

1. Create a user account with MFA active who can login to backend and frontend.
The user should have activated at least one MFA method which requires to enter a numeric verification code, e.g. TOTP, and another one which doesn't require such a code (Webauthn e.g with Windows Hello).
2. Login to the frontend with this account with username and password.
3. Check if depending on the MFA method, the input field for the verification code or the button to validate with the authenticator has focus.
Use the "Select a different method" button in the toolbar and check that for each method.
Watch the browser console with the developer tools of your browser to see if there are javascript errors.
Result: See section "Actual result BEFORE applying this Pull Request" without the PR, and "Expected result AFTER applying this Pull Request" with the PR.
4. Enter a verification code if necessary and use the validation button to finally log in.
5. Login to the backend with this account with username and password.
6. Check if depending on the MFA method, the input field for the verification code or the button to validate with the authenticator has focus.
Use the "Select a different method" button in the toolbar and check that for each method.
Watch the browser console with the developer tools of your browser to see if there are javascript errors.
Result: The MFA captive login page for the backend works with this PR on the same way as without this PR, i.e. nothing is broken by this PR, no javascript errors related to the "two-factor-focus" script in both cases.
7. Enter a verification code if necessary and use the validation button to finally log in.

### Actual result BEFORE applying this Pull Request

On the MFA captive login page for the frontend:
- When an MFA method is used which does not require to enter a numeric verification code, i.e. webauthn, no button has focus.
- When an MFA method is used which requires to enter a numeric verification code, e.g. TOTP or backup codes, the field to enter the verification code has focus, i.e. this works as expected.

On the MFA captive login page for the backend everything works as expected:
- When an MFA method is used which does not require to enter a numeric verification code, i.e. webauthn, the button to validate with the authenticator has focus.
- When an MFA method is used which requires to enter a numeric verification code, e.g. TOTP or backup codes, the field to enter the verification code has focus.

There are no javascript errors in the browser console related to the "two-factor-focus" script.

### Expected result AFTER applying this Pull Request

On the MFA captive login page for the frontend:
- When an MFA method is used which does not require to enter a numeric verification code, i.e. webauthn, the Submit button has focus.
- When an MFA method is used which requires to enter a numeric verification code, e.g. TOTP or backup codes, everything works as before applying this pull request, i.e. the field to enter the verification code has focus.

On the MFA captive login page for the backend: Everything works as well as before applying this pull request.

There are no javascript errors in the browser console related to the "two-factor-focus" script.

### Link to documentations
Please select:
- [X] No documentation changes for docs.joomla.org needed

- [X] No documentation changes for manual.joomla.org needed

richard67 committed:
Fix submit button focus on MFA captive site login

richard67 commented:
@nikosdion Hi Nik, could you check this PR and give a quick feedback? Maybe that difference between the admin and the site login captive pages has a reason of which I am just not aware?

dautrich commented:
I have tested this item :white_check_mark: successfully on 796d1458c4f14a3995b286f52f55e72409e5513f<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/40555">issues.joomla.org/tracker/joomla-cms/40555</a>.</sub>

nikosdion commented on build/media_source/com_users/js/two-factor-focus.es6.js:
```diff
@@ -17,21 +17,20 @@
-      if (elToolbarCustomElement) {
-        elToolbarCustomElement.querySelector('button').focus();
-      }
+    } else if (elToolbarCustomElement) {
+      elToolbarCustomElement.querySelector('button').focus();
+    } else if (elValidateButton) {
+      elValidateButton.focus();
```
No need to do that.

**You cannot focus a hidden element.**

Therefore the two consecutive `.focus()` will only ever focus the last of the two elements which is not hidden.

In other words, undo your change and go back to the code as it was before.

Here's an example for you to best understand what I wrote:

```html
<!DOCTYPE html>
<html>
<head>
	<title>Derp</title>
	<meta name="color-scheme" content="light dark">
	<style type="text/css">
        button.hidden {
            display: none
        }
	</style>
	<script type="text/javascript">
        window.addEventListener("DOMContentLoaded", (e) =>
        {
            document.querySelectorAll("button").forEach((el) =>
            {
               el.focus();

               el.addEventListener('click', (clickEvent) => {
                   clickEvent.preventDefault();

                   alert(clickEvent.target.innerText);
			   })
            });
        });
	</script>
</head>
<body>
<h1>Hello</h1>
<button class="hidden">Button 1</button>
<button>Button 2</button>
</body>
</html>
```

richard67 replied:
It does not do harm and saves us one check of an if condition if we do it my way.

richard67 replied:
And when doing only the change in the PHP file but not this change, it simply doesn't work. Try it yourself.

richard67 replied:
It would have been more helpful if you had answered my question why there is that difference between the admin and the site php files which I was asking about.

richard67 replied:
P.S. Maybe you should check ~~#40535~~ too to understand why it needs this change here now after that fix.

Sorry, I mean #40330 .

nikosdion replied:
No, what you showed me does not say that I am wrong. I insist that what you are doing here is **BAD** and will break something in the future.

Please keep in mind that Joomla's MFA did not come from nowhere. It was Akeeba LoginGuard, a product commercially available between 2016–2022 which I contributed to the Joomla! project. I understand what works and why. As for your question, I'll reply on the main thread.

richard67 replied:
@nikosdion Ok, thanks so far. Could you check the other PR #40330 , too? Should that be reverted?

nikosdion replied:
Not necessarily. I'd prefer using [optional chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining) instead, but that's a nitpick. I am neither a code absolutist nor a purist.

If you want to use optional chaining you can revert 40330 and replace your two changes with

```
const elToolbarCustomElement = document.getElementById('toolbar-user-mfa-submit')?.querySelector('button');
```

and

```
if (elToolbarCustomElement) {
    elToolbarCustomElement.focus();
}
```

The reason this works is that if there is no toolbar button the `document.getElementById('toolbar-user-mfa-submit')` returns a nullish value, therefore the optional chaining operator stops further chain execution and returns undefined, therefore `elToolbarCustomElement === undefined`.

As I said in the main thread, this code was hacked together at the eleventh hour. Another thing to note is that it was also the week I had people painting the outside of my house (it was going to be done, except there was a storm the week prior putting everything behind schedule), so I had _even less_ time to be in front of a screen. I tried avoiding that but, hey, it was not exactly under my control — I control neither what people demand to get a PR merged, nor the weather. Crap happens, I guess?

richard67 replied:
Thanks. As I'm not a JS guy and only try to fix things, I was not aware of the optional chaining. That's exactly what I was looking for. Thanks a lot for that.

richard67 replied:
Fixed.

nikosdion commented on build/media_source/com_users/js/two-factor-focus.es6.js:
```diff
@@ -17,21 +17,20 @@
         e.preventDefault();
 
-        elValidateButton.click();
+        if (elValidateButton) {
+          elValidateButton.click();
+        }
```
Why? What is the use case here? If we have a toolbar button we necessarily have an inline button, i.e. the if-block always executes because elValidateButton is always defined.

richard67 replied:
Not if someone has a layout override, and the if check doesn't do any harm.

richard67 replied:
@nikosdion It would be more helpful if you would answer my question why there is that difference between the admin and the site php files which I was asking about.

nikosdion replied:
I'll reply on the main thread.

richard67 replied:
I'll revert that later when I have time.

richard67 replied:
Done.

nikosdion commented:
@richard67 You asked me this:

> It would be more helpful if you would answer my question why there is that difference between the admin and the site php files which I was asking about.

I tried diplomatically avoid answering this because some people say in bad faith that I am difficult to work with. But if you insist, here's the whole story.

Between 2016 and 2022 I developed and made available free of charge my Akeeba LoginGuard extension for Joomla. This was the answer to everyone who was saying that captive MFA is impossible in Joomla — something I had been hearing since 2012 when I contributed the original TFA as a _temporary solution, not to be used for more than a year_, with the plan all along being captive MFA. I contributed this code last year as Joomla's MFA feature, after the project finally realised that the _temporary_ TFA solution was not all that secure, as I had warned a decade ago and dismissed ever since. Yay for common sense.

Akeeba LoginGuard used the exact same view templates in both the front- and the backend, and for good reasons: 1. a user MUST NOT be expected to log in in two _completely different_ ways depending on whether they are in the front- or the backend of the site; and 2. if the interface is not in the main content area users are confused and start looking around the page in desperation, not knowing what to do next.

However, my PR of a real world proven solution was not going to be accepted unless the backend MFA was made substantially different than the frontend MFA, moving the action button to the bloody toolbar. I vehemently disagree with this approach, it is proven to confuse users, and leads to lower adoption of MFA in Joomla. But I can not merge my own PRs. Therefore, I had to make two different view templates for the front- and backend. 

This was asked for in literally the eleventh hour so I hacked it together as best as I could. I didn't have time to do extensive testing, or we'd miss the Joomla! 4.2 merge window, the feature would languish in “PR Hell” and you'd never get MFA in Joomla… It was supposedly tested by the people who asked for this change and approved for merging by one of the maintainers. It does work, minus the focusing of the button in the frontend which is a regression to the original behaviour brought about by the hasty change in view templates and the JavaScript.

If you ask **me** what should be done, I will reiterate that the toolbar button is a really bad idea and must be removed. If it's removed, the JS can be simplified and the two view templates be the same. Exactly as it was for the 7 years people were using it under the name Akeeba LoginGuard.

If you do not want to remove the toolbar button we just need to look for both the toolbar button and the inline button by element ID, then run `.focus()` on them _if they are defined_ in the order inline, toolbar.

dautrich commented:
@nikosdion 
Can you have a look at issue  #40428, please. It's also about MFA login.

richard67 commented:
I've added a hint to the testing instructions about the ongoing review so people don't waste time with testing because the PR will receive changes. I haven't set it to draft because I don't want to keep people from reviewing. Thanks so far for all feedback, I've learned new things and will implement changes later today after my paid job.

richard67 committed:
Revert "[4.3] Fix javascript error on frontend MFA captive login page (#40330)"

This reverts commit b5f5db77106db005e8b3d499c90890a7a181a12a.

richard67 committed:
Use optional chaining operator

richard67 committed:
Merge branch '4.3-dev-fix-multifactorauth-captive-focus-for-site-3' into 4.3-dev-fix-multifactorauth-captive-focus-for-site-2

richard67 committed:
Revert unneccessary check for the button

richard67 committed:
PHPCS

richard67 commented:
@nikosdion I've implemented all your suggestions and updated the description accordingly. Thanks for the review.

@dautrich Could you redo your test with the latest changes? The testing instructions are still the same. The link to the update package and custom update URL is https://ci.joomla.org/artifacts/joomla/joomla-cms/4.3-dev/40555/downloads/65470/ . Thanks in advance.

dautrich commented:
I have tested this item :white_check_mark: successfully on 5d3abe5d399599667eefdecaaaa96bd0195466ed<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/40555">issues.joomla.org/tracker/joomla-cms/40555</a>.</sub>

richard67 committed:
Merge branch '4.3-dev' into 4.3-dev-fix-multifactorauth-captive-focus-for-site-2

obuisard commented:
I have tested this item :white_check_mark: successfully on 9474dcf6509736d27c1cd13ac392a03d0fa8a0e2<br /><br />Works as expected!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/40555">issues.joomla.org/tracker/joomla-cms/40555</a>.</sub>

obuisard committed:
Merge branch '4.3-dev' into 4.3-dev-fix-multifactorauth-captive-focus-for-site-2

obuisard committed:
Merge branch '4.3-dev' into 4.3-dev-fix-multifactorauth-captive-focus-for-site-2
